### PR TITLE
Adds support for directional relative permeabilities with hysteresis

### PR DESCRIPTION
**Pull request by hakonhagland**
Adds support for directional relative permeabilities with hysteresis (IMBNUMX, IMBNUMY, and IMBNUMZ).
To implement this, the current cell-wise material param vector (elements are of type `EclMultiplexerMaterialParams`) in `EclMaterialLawManager.hpp` is kept and three more material param vectors, one for each face X, Y, and Z are added. These vectors are allocated dynamically and used only if KRNUM{X,Y,Z} or IMBNUM{X,Y,Z} are specified in the input file.
The `initParamsForElements()` method in `EclMaterialLawManager.cpp` is refactored into different files and methods to make the implementation more readable and maintainable. 

A [test case](https://github.com/OPM/opm-tests/pull/884) has been added based on the existing test case `7_HYSTERESIS_MODEL2.DATA`.
Running this test case with this PR gives the following FOPR

Pasted image 20230116211834

See downstream https://github.com/OPM/opm-simulators/pull/4384 for the corresponding PR in opm-simulators

### CODE REVIEW

**Comment by hakonhagland** [comment]
jenkins build this opm-simulators=4384 please

**Comment by totto82** [comment]
The test indicates that there is something wrong with the gas+oil cases. The PR includes parts that seems like pure refactoring. I suggest that you compare the new and old code and look for differences related to gas+oil cases. The refactoring itself is good, so is the implementation of directional permeability with hysteresis. I.e. when Jenkins is happy this can be more or less merged as is.

**Comment by hakonhagland** [comment]
jenkins build this opm-simulators=4384 please

**Comment by OPMUSER** [comment]
I'm in the process I adding additional tests to the KRNUM tests would you like me to expand this to include IBNUM tests as well? The update will include comparative plots with the commercial simulator, with the included Markdown documentation.

**Comment by hakonhagland** [comment]
> The test indicates that there is something wrong with the gas+oil cases. ...  I suggest that you compare the new and old code and look for differences related to gas+oil cases.

@totto82 Good idea, I think I found the problem, see latest commit.

**Comment by totto82** [comment]
Thanks for the work @hakonhagland I have review the code and from my perspective this is fine. If not other comments I will merge this later today.

**Comment by totto82** [comment]
> I'm in the process I adding additional tests to the KRNUM tests would you like me to expand this to include IBNUM tests as well? The update will include comparative plots with the commercial simulator, with the included Markdown documentation.

That would be nice.

File: src/opm/material/fluidmatrixinteractions/EclMaterialLawManagerHystParams.cpp [hunk 1]
```diff
@@ -0,0 +1,291 @@
...
+namespace Opm {
+
+/* constructors*/
+template <class Traits>
+EclMaterialLawManager<Traits>::InitParams::HystParams::
+HystParams(EclMaterialLawManager<Traits>::InitParams& init_params) :
+    init_params_{init_params}, parent_{init_params_.parent_},
+    eclState_{init_params_.eclState_}
+{
+    gasOilParams_ = std::make_shared<GasOilTwoPhaseHystParams>();
```
Comment (GitPaean) [inline src/opm/material/fluidmatrixinteractions/EclMaterialLawManagerHystParams.cpp:34]: I can be wrong.  During the memory profiling, it looks like these three parameters can be pretty big.  For the SPE11C case, the three Params_  are allocated  59.2 MB each and about 180 MB in total.  And there is no HYSTER related in the SPE11C case. 

Maybe it is good only allocate memory for these when needed. @totto82

Comment (totto82) [reply src/opm/material/fluidmatrixinteractions/EclMaterialLawManagerHystParams.cpp:34]: The hystParams contains the two-phase params and end-point scaling etc so we can not remove it completely. But as I understand the code we currently allocate memory to the hysteresis parameters ++ even though we dont need them. Have you checked the memory profile if HYSTER is activated. Does it change significantly?

**Pull request merged**